File: tests/unit/python/test_aqrs_scale_overflow_guard.py
"""
tests/unit/python/test_aqrs_scale_overflow_guard.py
──────────────────────────────────────────────────────
Task #295 follow-up: raising scale_rank_max/additive_rank_max past the
old hardcoded 4 let a real fp8 MQAR curriculum run's per-channel
value_scale_k/output_scale_k grow unbounded (get_scale()'s combined
envelope S(row,col) has no clamp anywhere in delta_csr_types.hpp),
overflowing S in the forward pass and NaN-collapsing training (real
run, see conversation/JOURNAL.md). This file tests the fix: sili.
sparse_rnn._overflow_guard_array (the pure elementwise correction) and
DISLDOLayer/DISLDOLayer8.apply_scale_overflow_guard (the bulk C++
round-trip wiring). C++ coverage of the underlying bulk-vector
accessors already exists (test_aqrs_scale_overflow_guard.cpp) -- this
file is deliberately NOT re-deriving that; it confirms the Python-side
correction MATH and that it's actually reachable through a real layer.
"""
import numpy as np

from sili.sparse_rnn import DISLDOLayer, DISLDOLayer8, _overflow_guard_array


class TestOverflowGuardArrayMath:
    def test_values_within_near_pass_through_unchanged(self):
        x = np.array([-5.0, -1.0, 0.0, 3.0, 10.0], dtype=np.float32)
        out = _overflow_guard_array(x, clip=200.0, near=20.0, coef=0.1)
        np.testing.assert_allclose(out, x, atol=1e-5)

    def test_values_past_near_get_shrunk_toward_zero(self):
        # near=20, so a value of 50 has excess=30 -- corrected should be
        # strictly closer to zero than the original, and same sign.
        x = np.array([50.0, -50.0], dtype=np.float32)
        out = _overflow_guard_array(x, clip=200.0, near=20.0, coef=0.1)
        assert abs(out[0]) < abs(x[0])
        assert abs(out[1]) < abs(x[1])
        assert out[0] > 0 and out[1] < 0  # sign preserved, not flipped

    def test_hard_clip_never_exceeded(self):
        x = np.array([1e9, -1e9, 5e6], dtype=np.float32)
        out = _overflow_guard_array(x, clip=200.0, near=20.0, coef=0.1)
        assert np.all(np.abs(out) <= 200.0 + 1e-3)

    def test_nan_and_inf_are_sanitized_not_passed_through(self):
        # The real bug: np.clip(nan, ...) == nan, so a plain clip alone
        # would NOT fix an already-NaN value_scale_k -- confirms the
        # explicit nan_to_num step actually does something.
        x = np.array([np.nan, np.inf, -np.inf], dtype=np.float32)
        out = _overflow_guard_array(x, clip=200.0, near=20.0, coef=0.1)
        assert np.all(np.isfinite(out))
        assert out[1] > 0  # +inf -> +clip (sign preserved)
        assert out[2] < 0  # -inf -> -clip

    def test_near_at_or_below_zero_is_a_pure_hard_clip(self):
        # near=0 degenerates to "always apply correction" -- sanity
        # check the formula doesn't blow up at the boundary.
        x = np.array([0.5, -0.5], dtype=np.float32)
        out = _overflow_guard_array(x, clip=200.0, near=0.0, coef=0.1)
        assert np.all(np.isfinite(out))


def _make_dense_layer(cls, n_in=8, n_out=8, budget=2000, cpus=1, seed=0):
    rng = np.random.default_rng(seed)
    return cls(n_in, n_out, budget, cpus, rng=rng, dense=True)


class TestApplyScaleOverflowGuardIntegration:
    def test_fp4_layer_shrinks_artificially_inflated_value_scale(self):
        layer = _make_dense_layer(DISLDOLayer)
        layer._c.set_scale_rank_max(4)
        layer._c.set_scale_rank(4)
        # Directly inflate one channel to a dangerous magnitude, as if
        # several RMSprop steps had driven it there.
        layer._c.set_value_scale_raw_k(0, 1, 5000.0)
        before = layer._c.get_value_scale_k(0, 1)
        assert before == 5000.0

        layer.apply_scale_overflow_guard(clip=200.0, near=20.0, coef=0.1)
        after = layer._c.get_value_scale_k(0, 1)
        assert after < before
        assert after <= 200.0 + 1e-3
        assert np.isfinite(after)

    def test_fp4_layer_untouched_channel_stays_within_near(self):
        layer = _make_dense_layer(DISLDOLayer)
        layer._c.set_scale_rank_max(4)
        layer._c.set_scale_rank(4)
        before = layer._c.get_value_scale_k(0, 0)  # component 0 default 1.0
        layer.apply_scale_overflow_guard(clip=200.0, near=20.0, coef=0.1)
        after = layer._c.get_value_scale_k(0, 0)
        assert after == before  # well within `near`, untouched

    def test_fp4_layer_rank0_additive_is_a_safe_noop(self):
        # additive_rank defaults to 0 -- get_additive_u_raw_vector()
        # should be empty, and apply_scale_overflow_guard must not
        # error on an empty array (see _apply_scale_overflow_guard's
        # own `if au.size:` guard).
        layer = _make_dense_layer(DISLDOLayer)
        assert len(layer._c.get_additive_u_raw_vector()) == 0
        layer.apply_scale_overflow_guard()  # must not raise

    def test_fp8_layer_shrinks_artificially_inflated_output_scale(self):
        layer = _make_dense_layer(DISLDOLayer8)
        layer._c.set_scale_rank_max(4)
        layer._c.set_scale_rank(4)
        layer._c.set_output_scale_raw_k(0, 2, -8000.0)
        before = layer._c.get_output_scale_k(0, 2)
        assert before == -8000.0

        layer.apply_scale_overflow_guard(clip=200.0, near=20.0, coef=0.1)
        after = layer._c.get_output_scale_k(0, 2)
        assert after > before  # moved toward zero (less negative)
        assert after >= -200.0 - 1e-3
        assert np.isfinite(after)

    def test_fp4_layer_additive_channel_gets_corrected_when_present(self):
        layer = _make_dense_layer(DISLDOLayer)
        layer._c.set_additive_rank_max(4)
        layer._c.set_additive_rank(4)
        layer._c.set_additive_u_raw_k(0, 1, 3000.0)
        before = layer._c.get_additive_u_k(0, 1)
        layer.apply_scale_overflow_guard(clip=200.0, near=20.0, coef=0.1)
        after = layer._c.get_additive_u_k(0, 1)
        assert after < before
        assert np.isfinite(after)

    def test_repeated_calls_converge_not_diverge(self):
        # Applying the guard repeatedly (as it would be every training
        # step) must monotonically shrink an over-threshold value, not
        # oscillate or grow -- a real correctness property of the
        # auto-correcting design, not just a one-shot check.
        layer = _make_dense_layer(DISLDOLayer)
        layer._c.set_scale_rank_max(4)
        layer._c.set_scale_rank(4)
        layer._c.set_value_scale_raw_k(0, 1, 1000.0)
        prev = 1000.0
        for _ in range(20):
            layer.apply_scale_overflow_guard(clip=200.0, near=20.0, coef=0.1)
            cur = layer._c.get_value_scale_k(0, 1)
            assert cur <= prev + 1e-6
            prev = cur
        assert prev < 1000.0

File: tests/unit/test_aqrs_scale_overflow_guard.cpp
// AQRS scale/additive channel bulk raw-vector accessors (task #295
// follow-up / task #286 "virtual neuron" exposure): get/set_value_scale_
// raw_vector, get/set_output_scale_raw_vector, get/set_additive_u_raw_
// vector, get/set_additive_v_raw_vector on SparseLinearWeightsDelta
// (delta_csr_types.hpp). Built to let a caller (Python) read/correct the
// AQRS scale envelope in ONE call instead of n*rank individual accessor
// calls -- the real motivation: a raised scale_rank_max/additive_rank_max
// (task #295) let a real fp8 MQAR curriculum run's per-channel
// value_scale_k/output_scale_k grow unbounded (get_scale()'s combined
// envelope has no clamp anywhere), overflowing S in the forward pass and
// NaN-collapsing training. This test proves the exposure primitive
// itself is correct (round-trips, size-preserving); the actual
// clip+auto-correct policy lives in Python (sili_peridot's own
// _overflow_guard_array, sili/sparse_rnn.py).
#include "../../sili/lib/headers/delta_csr_memory.hpp"
#include <cstdio>
#include <cmath>
#include <vector>

static int g_fail = 0;
#define CHECK(cond, fmt, ...)                                                                      \
    do {                                                                                           \
        if (!(cond)) {                                                                             \
            std::printf("FAIL %s:%d: " fmt "\n", __FILE__, __LINE__, ##__VA_ARGS__);               \
            std::fflush(stdout);                                                                   \
            ++g_fail;                                                                              \
        }                                                                                          \
    } while (0)

using SIZE_TYPE = int;
using COL_TYPE = uint32_t;
using Weights = SparseLinearWeightsDelta<SIZE_TYPE, FP4BiPacked, COL_TYPE>;

static const std::size_t N = 8;

static Weights make_layer() {
    Weights w;
    std::vector<SIZE_TYPE> ptrs(N + 1);
    std::vector<SIZE_TYPE> idx(N * N);
    std::vector<float> vals(N * N, 1.0f), imp(N * N, 1.0f);
    for (std::size_t r = 0; r < N; ++r) {
        ptrs[r] = SIZE_TYPE(r * N);
        for (std::size_t c = 0; c < N; ++c)
            idx[r * N + c] = SIZE_TYPE(c);
    }
    ptrs[N] = SIZE_TYPE(N * N);
    w.connections = delta_csr_from_absolute<SIZE_TYPE, FP4BiPacked, COL_TYPE>(
        ptrs, idx, vals, imp, N, N, N * N * 2, N * N * 2);
    w.out_degree.assign(N, SIZE_TYPE(N));
    return w;
}

// Bulk value_scale/output_scale vectors are empty until the first
// set_*_raw_k touches them (lazy sizing, matches every other AQRS
// storage member's own convention).
static void test_empty_before_any_write() {
    Weights w = make_layer();
    CHECK(w.get_value_scale_raw_vector().empty(), "value_scale should start empty");
    CHECK(w.get_output_scale_raw_vector().empty(), "output_scale should start empty");
    CHECK(w.get_additive_u_raw_vector().empty(), "additive_u should start empty");
    CHECK(w.get_additive_v_raw_vector().empty(), "additive_v should start empty");
}

// Round-trip: bulk-get, mutate, bulk-set, confirm per-element accessors
// see the SAME values -- proves the bulk vector really is the live
// backing storage (a copy that silently diverged would fail this).
static void test_value_scale_round_trip() {
    Weights w = make_layer();
    w.set_scale_rank_max(4);
    w.set_scale_rank(4);
    for (std::size_t r = 0; r < N; ++r)
        for (std::size_t k = 0; k < 4; ++k)
            w.set_value_scale_raw_k(r, k, float(r * 10 + k));

    auto vs = w.get_value_scale_raw_vector();
    CHECK(vs.size() == N * 4, "expected value_scale vector size %zu, got %zu", N * 4, vs.size());
    for (auto& x : vs)
        x *= 2.0f;
    w.set_value_scale_raw_vector(vs);

    for (std::size_t r = 0; r < N; ++r)
        for (std::size_t k = 0; k < 4; ++k) {
            float expected = float(r * 10 + k) * 2.0f;
            float got = w.get_value_scale_k(r, k);
            CHECK(std::fabs(got - expected) < 1e-4f,
                  "row=%zu k=%zu: expected %.3f after bulk round-trip, got %.3f", r, k, expected,
                  got);
        }
}

static void test_output_scale_round_trip() {
    Weights w = make_layer();
    w.set_scale_rank_max(3);
    w.set_scale_rank(3);
    for (std::size_t c = 0; c < N; ++c)
        for (std::size_t k = 0; k < 3; ++k)
            w.set_output_scale_raw_k(c, k, float(c + k) * 0.5f);

    auto os = w.get_output_scale_raw_vector();
    CHECK(os.size() == N * 3, "expected output_scale vector size %zu, got %zu", N * 3, os.size());
    for (auto& x : os)
        x += 100.0f;
    w.set_output_scale_raw_vector(os);

    for (std::size_t c = 0; c < N; ++c)
        for (std::size_t k = 0; k < 3; ++k) {
            float expected = float(c + k) * 0.5f + 100.0f;
            float got = w.get_output_scale_k(c, k);
            CHECK(std::fabs(got - expected) < 1e-4f,
                  "col=%zu k=%zu: expected %.3f after bulk round-trip, got %.3f", c, k, expected,
                  got);
        }
}

static void test_additive_u_v_round_trip() {
    Weights w = make_layer();
    w.set_additive_rank_max(5);
    w.set_additive_rank(5);
    for (std::size_t r = 0; r < N; ++r)
        for (std::size_t k = 0; k < 5; ++k)
            w.set_additive_u_raw_k(r, k, float(r) - float(k));
    for (std::size_t c = 0; c < N; ++c)
        for (std::size_t k = 0; k < 5; ++k)
            w.set_additive_v_raw_k(c, k, float(c) + float(k));

    auto au = w.get_additive_u_raw_vector();
    auto av = w.get_additive_v_raw_vector();
    CHECK(au.size() == N * 5, "expected additive_u size %zu, got %zu", N * 5, au.size());
    CHECK(av.size() == N * 5, "expected additive_v size %zu, got %zu", N * 5, av.size());
    for (auto& x : au)
        x = -x;
    for (auto& x : av)
        x = -x;
    w.set_additive_u_raw_vector(au);
    w.set_additive_v_raw_vector(av);

    for (std::size_t r = 0; r < N; ++r)
        for (std::size_t k = 0; k < 5; ++k) {
            float expected = -(float(r) - float(k));
            CHECK(std::fabs(w.get_additive_u_k(r, k) - expected) < 1e-4f,
                  "additive_u row=%zu k=%zu mismatch after bulk round-trip", r, k);
        }
    for (std::size_t c = 0; c < N; ++c)
        for (std::size_t k = 0; k < 5; ++k) {
            float expected = -(float(c) + float(k));
            CHECK(std::fabs(w.get_additive_v_k(c, k) - expected) < 1e-4f,
                  "additive_v col=%zu k=%zu mismatch after bulk round-trip", c, k);
        }
}

// Size-preserving: a bulk setter must reject anything that isn't
// EXACTLY the current size -- this is a correction pass over existing
// trained values, not a resize/grow path (that's set_scale_rank's job,
// and a silent length change here would corrupt the row/col-major
// layout every _k accessor depends on).
static void test_size_mismatch_throws() {
    Weights w = make_layer();
    w.set_scale_rank_max(4);
    w.set_scale_rank(4);
    w.set_value_scale_raw_k(0, 0, 1.0f); // touch it so the vector is non-empty
    bool threw = false;
    try {
        w.set_value_scale_raw_vector(std::vector<float>{1.0f, 2.0f});
    } catch (const std::invalid_argument&) {
        threw = true;
    }
    CHECK(threw, "set_value_scale_raw_vector should throw on size mismatch");

    threw = false;
    try {
        w.set_output_scale_raw_vector(std::vector<float>(1000, 0.0f));
    } catch (const std::invalid_argument&) {
        threw = true;
    }
    CHECK(threw, "set_output_scale_raw_vector should throw on size mismatch");

    w.set_additive_rank_max(2);
    w.set_additive_rank(2);
    w.set_additive_u_raw_k(0, 0, 1.0f);
    threw = false;
    try {
        w.set_additive_u_raw_vector(std::vector<float>{});
    } catch (const std::invalid_argument&) {
        threw = true;
    }
    CHECK(threw, "set_additive_u_raw_vector should throw on size mismatch (empty vs non-empty)");
}

// Exact-size (no-op content) set must NOT throw -- confirms the check
// is size-only, not some other spurious rejection.
static void test_exact_size_set_does_not_throw() {
    Weights w = make_layer();
    w.set_scale_rank_max(2);
    w.set_scale_rank(2);
    w.set_value_scale_raw_k(0, 0, 1.0f);
    auto vs = w.get_value_scale_raw_vector();
    bool threw = false;
    try {
        w.set_value_scale_raw_vector(vs);
    } catch (const std::invalid_argument&) {
        threw = true;
    }
    CHECK(!threw, "set_value_scale_raw_vector should NOT throw on exact-size input");
}

int main() {
    test_empty_before_any_write();
    test_value_scale_round_trip();
    test_output_scale_round_trip();
    test_additive_u_v_round_trip();
    test_size_mismatch_throws();
    test_exact_size_set_does_not_throw();

    if (g_fail == 0) {
        std::printf("All AQRS scale overflow guard exposure tests passed.\n");
    } else {
        std::printf("%d FAILURES\n", g_fail);
    }
    return g_fail == 0 ? 0 : 1;
}
